method/SEDR/SEDR_method.py

```diff
@@ -0,0 +1,313 @@
+)
+parser.add_argument(
+    "--config",
+    help="Optional config file (json) used to pass additional parameters.",
+    required=False,
+)
+
+## Session for code
+args = parser.parse_args()
+
+# anndata input
+def get_anndata(args):
+    import anndata as ad
+    import numpy as np
+    import pandas as pd
+    import scipy as sp
+
+    observations = pd.read_table(args.observations, index_col=0)
+    features = pd.read_table(args.features, index_col=0)
+
+    coordinates = (
+        pd.read_table(args.coordinates, index_col=0)
+        .loc[observations.index, :]
+        .to_numpy()
+    )
+
+    adata = ad.AnnData(obs=observations, var=features, obsm={"spatial": coordinates})
+
+    if args.matrix is not None:
+        X = sp.io.mmread(args.matrix)
+        if sp.sparse.issparse(X):
+            X = X.tocsr()
+        adata.X = X
+
+    if args.neighbors is not None:
+        adata.obsp["spatial_connectivities"] = sp.io.mmread(args.neighbors).T.tocsr()
+
+    # Filter by selected samples/features
+    if "selected" in adata.obs.columns:
+        adata = adata[observations["selected"].astype(bool), :]
+    if "selected" in adata.var.columns:
+        adata = adata[:, features["selected"].astype(bool)]
+
+    if args.dim_red is not None:
+        adata.obsm["reduced_dimensions"] = (
+            pd.read_table(args.dim_red, index_col=0).loc[adata.obs_names].to_numpy()
+        )
+
+    return adata
+
+## Cluster code is SEDR had some bugs. Modified below
+def res_search_fixed_clus_leiden(adata, n_clusters, increment=0.01, random_seed=2023):
+    import scanpy as sc
+    import pandas as pd
+    import numpy as np
+    
+    for res in np.arange(0.2, 2, increment):
+        sc.tl.leiden(adata, random_state=random_seed, resolution=res)
+        if len(adata.obs['leiden'].unique()) > n_clusters:
+            break
+    return res-increment
+
+
+def leiden(adata, n_clusters, use_rep='SEDR', key_added='SEDR', random_seed=2023):
+    import scanpy as sc
+    import pandas as pd
+    import numpy as np
+    
+    sc.pp.neighbors(adata, use_rep=use_rep)
+    res = res_search_fixed_clus_leiden(adata, n_clusters, increment=0.01, random_seed=random_seed)
+    sc.tl.leiden(adata, random_state=random_seed, resolution=res)
+
+    adata.obs[key_added] = adata.obs['leiden']
+    adata.obs[key_added] = adata.obs[key_added].astype('int')
+    adata.obs[key_added] = adata.obs[key_added].astype('category')
+
+    return adata
+
+
+def res_search_fixed_clus_louvain(adata, n_clusters, increment=0.01, random_seed=2023):
+    import scanpy as sc
+    import pandas as pd
+    import numpy as np
+    
+    for res in np.arange(0.2, 2, increment):
+        sc.tl.louvain(adata, random_state=random_seed, resolution=res)
+        if len(adata.obs['louvain'].unique()) > n_clusters:
+            break
+    return res-increment
+
+def louvain(adata, n_clusters, use_rep='SEDR', key_added='SEDR', random_seed=2023):
```

Comment:
please use the provided seeds (in all instances in this file)

Reply:
Hi, yes this is only defined as a last-resort default in case there's no seed provided by the user input. In the final clustering functions, we use all the seeds defined by the input. (see line 279-301). I can also remove the default if it's the most convenient for you. :)
```
match config['cluster_method']:
    case "mclust":
        adata = mclust_R(adata, 
                              n_clusters = n_clusters, 
                              use_rep='SEDR', 
                              key_added='SEDR', 
                              random_seed=seed    #user-defined seeds
                             )
    case "louvain":
         adata = louvain(adata, 
                              n_clusters = n_clusters, 
                              use_rep='SEDR', 
                              key_added='SEDR', 
                              random_seed=seed
                             )
    case "leiden":
         adata = leiden(adata, 
                              n_clusters = n_clusters, 
                              use_rep='SEDR', 
                              key_added='SEDR', 
                              random_seed=seed
                             )
```

Reply:
Sorry missed that. Should be fine then.